src/server/generic_family.cc

```diff
@@ -259,6 +465,39 @@ OpResult<std::string> OpDump(const OpArgs& op_args, string_view key) {
   return OpStatus::KEY_NOTFOUND;
 }
 
+OpResult<bool> OnRestore(const OpArgs& op_args, std::string_view key, std::string_view payload,
+                         RestoreArgs restore_args) {
+  if (!restore_args.UpdateExpiration(op_args.db_cntx.time_now_ms)) {
+    return OpStatus::OUT_OF_RANGE;
+  }
+
+  auto& db_slice = op_args.shard->db_slice();
+  // The redis impl (see cluster.c function restoreCommand), remove the old key if
+  // the replace option is set, so lets do the same here
+  auto [from_it, from_expire] = db_slice.FindExt(op_args.db_cntx, key);
+  if (restore_args.Replace()) {
+    if (IsValid(from_it)) {
+      VLOG(1) << "restore command is running with replace, found old key '" << key
+              << "' and removing it";
+      CHECK(db_slice.Del(op_args.db_cntx.db_index, from_it));
+    }
+  } else {
+    // we are not allowed to replace it, so make sure it doesn't exist
+    if (IsValid(from_it)) {
+      return OpStatus::KEY_EXISTS;
+    }
+  }
+
+  if (restore_args.Expired()) {
+    VLOG(1) << "the new key '" << key << "' already expired, will not save the value";
+    return true;
+  }
+
+  RdbRestoreValue loader(payload);
+  return loader.Read() &&
```

Comment:
1. You can just pass payload as an argument to Read(). Why do you need a stateful object for that?
2.  I would rename `Read()` to `Parse(string_view)` and similarly this action can be done outside inside `Restore` function.

Reply:
will remove this class - we can live without it, you're right

Reply:
Actually it wasn't the class to remove only the state for this derived class